### PR TITLE
stubtest: Reduce false-positive errors on runtime type aliases

### LINKED ISSUE #12821: stubtest false positive with runtime type alias for Callable
From `python -m mypy.stubtest --custom-typeshed-dir . wsgiref`:

```
error: wsgiref.types.WSGIApplication is not a recognised type alias
Stub: at line 14
def (builtins.dict[builtins.str, Any], wsgiref.types.StartResponse) -> typing.Iterable[builtins.bytes]
Runtime: in file /Users/jelle/py/cpython/Lib/collections/abc.py
collections.abc.Callable[[dict[str, typing.Any], wsgiref.types.StartResponse], collections.abc.Iterable[bytes]]
```

Once again, the stub and implementation are exactly the same. It's odd that the stub infers a `def (...)` while the runtime infers a `collections.abc.Callable`.

### LINKED ISSUE #13114: stubtest: Problem with runtime type aliases
**Bug Report**

Using mypy 0.960 and Python 3.9.2, stubtest reports a mismatch between runtime and a stub file when both use an identical `TypeAlias` annotation:

```
error: wsgiref.types.WSGIEnvironment is not a type
Stub: at line 1003
<TypeInfo builtins.dict>
Runtime:
dict[str, typing.Any]
```

(Source: https://github.com/python/typeshed/runs/7319247103?check_suite_focus=true)

In this case, the offending line is `WSGIEnvironment: TypeAlias = dict[str, Any]`, both at runtime and the stub. Interestingly, the following line works ok:

```python
WSGIApplication: TypeAlias = Callable[[WSGIEnvironment, StartResponse], Iterable[bytes]]
```

### PR DESCRIPTION
**Pull request by AlexWaygood**
### Description

Stubtest doesn't do very well at dealing with type aliases that exist at runtime. This PR fixes that.

Fixes #13114. Fixes #12821.

## Test Plan

Several tests added.

### CODE REVIEW

**Comment by AlexWaygood** [comment]
This PR depends on #13101, since it uses `typing_extensions.get_origin`, which only works on 3.7+.

**Review by hauntsaninja — COMMENTED** [review]
Thanks, this looks pretty good. I made a couple minor changes.

One quick thing: maybe it's worth checking full names? It does seem like the most common issue here would be cases where the stub has the name right but the qualified name wrong. We could do a quick `{"typing": "collections.abc"}.get(mod_name, mod_name)` to normalise.

**Comment by AlexWaygood** [comment]
> Thanks, this looks pretty good. I made a couple minor changes.

Thanks for the cleanups — they all look like changes for the better 😊



> One quick thing: maybe it's worth checking full names? It does seem like the most common issue here would be cases where the stub has the name right but the qualified name wrong. We could do a quick `{"typing": "collections.abc"}.get(mod_name, mod_name)` to normalise.

Yeah, if it's just `typing`/`collections.abc`, it's not much of a problem. But what about e.g. `StringIO`,  which is in `io.pyi` in typeshed, but has `_io` as its `__module__` at runtime. And what if somebody's using a Python implementation where the stdlib `io` module imports `StringIO` from `_pyio` instead of `_io`? We'd want an object by the runtime name of `_io.StringIO` or `_pyio.StringIO` to be compatible with typeshed's `io.StringIO`. This principle could apply to an arbitrary number of hypothetical third-party modules, as well as to all the modules in the stdlib with C-accelerator modules.

**Review by hauntsaninja — COMMENTED** [review]
Right. I remember running into some of this stuff when I tried to get stubtest to check MRO. I had something that had a couple heuristics and worked okay, but it was icky enough that I never merged it.

That said, all hope is not lost. We could try the fullname match (adjusting for typing / collections.abc). And if it's not a match, we can fallback to the current behaviour of the structural check `yield from verify(stub_target.type, runtime, object_path)`.

It's a nice optimisation over master, avoids false positives from typing constructs, avoids duplicate errors in the common case, avoids false positives for cases where the runtime itself gets aliased (`class _A: ...; A = _A; A.__name__ = _A`), keeps true positives for bad fully qualified aliases.

File: mypy/stubtest.py [hunk 1]
```diff
@@ -1010,33 +1011,70 @@ def verify_typealias(
...
+            runtime_name = getattr(runtime_origin, "__name__", MISSING)
+        if isinstance(runtime_name, str):
+            runtime_module: MaybeMissing[str] = getattr(runtime_origin, "__module__", MISSING)
+            if isinstance(runtime_module, str):
+                if runtime_module == "collections.abc" or (
+                    runtime_module == "re" and runtime_name in {"Match", "Pattern"}
+                ):
+                    runtime_module = "typing"
+                runtime_fullname = f"{runtime_module}.{runtime_name}"
+                if re.match(fr"_?{stub_target.type.fullname}", runtime_fullname):
```
Comment (AlexWaygood) [inline mypy/stubtest.py:1044]:
```suggestion
                if re.match(fr"_?{re.escape(stub_target.type.fullname)}", runtime_fullname):
```

**Comment by AlexWaygood** [comment]
Here's the output if I run typeshed's `stubtest_stdlib.py` with this patch checked out:

```diff
+note: unused allowlist entry sqlite3.dbapi2.Binary.__contains__
+note: unused allowlist entry sqlite3.Binary.__contains__
+note: unused allowlist entry _socket.error.characters_written
+note: unused allowlist entry builtins.EnvironmentError.characters_written
+note: unused allowlist entry builtins.IOError.characters_written
+note: unused allowlist entry dbm.dumb.error.characters_written
+note: unused allowlist entry os.error.characters_written
+note: unused allowlist entry select.error.characters_written
+note: unused allowlist entry socket.error.characters_written
+note: unused allowlist entry builtins.WindowsError.characters_written
+note: unused allowlist entry winreg.error.characters_written
```

This is good. `_socket.error`, `builtins.EnvironmentError`, `builtins.IOError`, `dbm.dumb.error`, `os.error`, `select.error`, `socket.error`, `builtins.WindowsError` and `winreg.error` are all aliases to `builtins.OSError`. `sqlite3.dbapi2.Binary` and `sqlite3.Binary` are both aliases to `builtins.memoryview`.

There's no need for stubtest to emit duplicate errors for all of these aliases, given that it already emits an error for `OS.characters_written` and `memoryview.__contains__`. With this patch, stubtest is able to detect that the fullnames are the same, and can therefore pass on emitting the duplicate errors.

File: mypy/stubtest.py [hunk 2]
```diff
@@ -1010,33 +1011,71 @@ def verify_typealias(
...
+            runtime_name = getattr(runtime_origin, "__name__", MISSING)
+        if isinstance(runtime_name, str):
+            runtime_module: object = getattr(runtime_origin, "__module__", MISSING)
+            if isinstance(runtime_module, str):
+                if runtime_module == "collections.abc" or (
+                    runtime_module == "re" and runtime_name in {"Match", "Pattern"}
+                ):
+                    runtime_module = "typing"
+                runtime_fullname = f"{runtime_module}.{runtime_name}"
+                if re.match(fr"_?{re.escape(stub_origin.fullname)}", runtime_fullname):
```
Comment (hauntsaninja) [inline mypy/stubtest.py:1045]: Should be fullmatch? Or is that intentional?

Comment (AlexWaygood) [reply mypy/stubtest.py:1045]: Not intentional — fullmatch is better here, yup!

```suggestion
                if re.fullmatch(fr"_?{re.escape(stub_origin.fullname)}", runtime_fullname):
```

**Review by hauntsaninja — COMMENTED** [review]
Yay all green! One last nit, do we know for sure that `runtime.__module__ + runtime.__qualname__ == stub.fullname`? I'm realising I'm not actually sure what `stub.fullname` is for all the nested cases.

**Comment by AlexWaygood** [comment]
> Yay all green! One last nit, do we know for sure that `runtime.__module__ + runtime.__qualname__ == stub.fullname`? I'm realising I'm not actually sure what `stub.fullname` is for all the nested cases.

I applied this diff to `mypy/stubtest.py` locally:

```diff
diff --git a/mypy/stubtest.py b/mypy/stubtest.py
index 96f6aa5af..1d65a6af1 100644
--- a/mypy/stubtest.py
+++ b/mypy/stubtest.py
@@ -1055,6 +1055,8 @@ def verify_typealias(
             runtime_name = runtime_origin.__qualname__
         except AttributeError:
             runtime_name = getattr(runtime_origin, "__name__", MISSING)
+        if runtime_origin.__qualname__ != runtime_origin.__name__:
+            print(f'{runtime_origin.__module__}.{runtime_origin.__qualname__}: {stub_origin.fullname}')
         if isinstance(runtime_name, str):
             runtime_module: object = getattr(runtime_origin, "__module__", MISSING)
             if isinstance(runtime_module, str):
```

And then ran `pytest ./mypy/test/teststubtest.py`. The printed output was:

```
test_module.Foo.Bar: test_module.Foo.Bar
```

So I think we're good!

**Comment by hauntsaninja** [comment]
Hooray! Thank you

**Comment by AlexWaygood** [comment]
Wooooo 😃🥳

**Pull request merged**